utils/self-extr-exec/compressor.cpp

```diff
@@ -0,0 +1,352 @@
+            delete [] files_data;
+            return 1;
+        }
+
+        /// Remember information about file name
+        /// This should be made after the file is opened
+        /// because filename should be extracted from path
+        names[i] = strrchr(filenames[i], '/') + 1;
+        files_data[i].name_length = strlen(names[i]);
+        sum_file_size += files_data[i].name_length;
+
+        /// read data about input file
+        struct stat info_in;
+        if (0 != fstat(input_fd, &info_in))
+        {
+            perror(nullptr);
+            delete [] files_data;
+            return 1;
+        }
+
+        if (info_in.st_size == 0)
+        {
+            std::cout << "Empty input file will be skipped." << std::endl;
+            continue;
+        }
+
+        std::cout << "Input file current size is " << info_in.st_size << std::endl;
+
+        /// Remember information about uncompressed size of file and
+        /// start of it's compression version
+        files_data[i].uncompressed_size = info_in.st_size;
+        files_data[i].start = pointer;
+
+        /// Compressed data will be added to the end of file
+        /// It will allow to create self extracting executable from file
+        if (0 != compress(input_fd, output_fd, level, pointer, info_in))
+        {
+            perror(nullptr);
+            delete [] files_data;
+            return 1;
+        }
+
+        /// This error is less important, than others.
+        /// If file cannot be closed, in some cases it will lead to
+        /// error in other function that will stop compression process
+        if (0 != close(input_fd))
+            perror(nullptr);
+
+        files_data[i].end = pointer;
+    }
+
+    /// save location of files information
+    metadata.start_of_files_data = pointer;
+
+    if (0 != saveMetaData(names, count, output_fd, metadata, files_data, pointer, sum_file_size))
+    {
+        delete [] files_data;
+        return 1;
+    }
+
+    delete [] files_data;
+    return 0;
+}
+
+int main(int argc, char* argv[])
+{
+    if (argc < 3)
+    {
+        std::cout << "Not enough arguments.\ncompressor [OPTIONAL --level of compression] [file name for compressed file] [files that should be compressed]" << std::endl;
+        return 0;
+    }
+
+    int start_of_files = 1;
+
+    /// Set compression level
+    int level = 5;
+    if (0 == memcmp(argv[1], "--level=", 8))
+    {
+        level = strtol(argv[1] + 8, nullptr, 10);
+        ++start_of_files;
+    }
+
+    int output_fd = open(argv[start_of_files], O_RDWR | O_CREAT, 0775);
+    if (output_fd == -1)
+    {
+        perror(nullptr);
+        return 1;
+    }
+    ++start_of_files;
+
+    struct stat info_out;
+    if (0 != fstat(output_fd, &info_out))
+    {
+        perror(nullptr);
+        return 1;
+    }
+
+    std::cout << "Compression with level " << level << std::endl;
+    if (0 != compressFiles(&argv[start_of_files], argc - start_of_files, output_fd, level, info_out))
```

Comment:
As I understood from decompressor (and from Overview of compression) - compressed file should start with decompressor executable? I don't see where it included...